listenbrainz_spark/recommendations/candidate_sets.py

```diff
@@ -0,0 +1,363 @@
+    """ Prepare similar artists dataframe which consists of top x (limit) artists similar
+        to the top artist given as argument where x is a config value. The function is invoked
+        when candidate set HTML is to be generated.
+
+        Args:
+            artist (str): Name of one of the top artist.
+
+        Returns:
+            similar_artists_df (dataframe): Dataframe consisting of artists similar to the top artist
+                given as argument with limit x. Dataframe column can be depicted as:
+                    ['artist_name_1']
+    """
+    similar_artists_df = run_query("""
+        SELECT artist_name_1
+            FROM artists_relation
+         WHERE artist_name_0 = "%s"
+        ORDER BY count DESC
+        LIMIT %s
+    """ % (artist, config.SIMILAR_ARTISTS_LIMIT))
+    return similar_artists_df
+
+def get_candidate_html_data(top_artist_with_collab, user_name):
+    """ Get artists similar to top artists listened to by the user. The function is invoked
+        when candidate set HTML is to be generated.
+
+        Args:
+            top_artists_with_collab (dataframe): Dataframe of top artists listened to by the user
+                whose similar artists count is not zero.
+            user_name (str): User name of the user.
+
+        Returns:
+            artists (dict): Dictionary can be depicted as:
+                {
+                    'top_artists 1' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                    'top_artists 2' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                    .
+                    .
+                    .
+                    'top_artists y' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                }
+    """
+    artists = defaultdict(dict)
+    for row in top_artist_with_collab.collect():
+        similar_artists_df = get_similar_artist_candidate_html(row.artist_name_0)
+        artists[row.artist_name_0] = [row.artist_name_1 for row in similar_artists_df.collect()]
+    return artists
+
+def save_candidate_html(user_data):
+    """ Save user data to an HTML file.
+
+        Args:
+            user_data (dict): Dcitionary can be depicted as:
+                {
+                    'user_name 1': {
+                        'artists': {
+                            'top_artists 1' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                        ...
+                        'top_artists y' : ['similar_artist 1', 'simialr_artist 2' ... 'similar_artist x'],
+                        },
+                        'time' : 'xxx'
+                    },
+                }
+    """
+    date = datetime.utcnow().strftime('%Y-%m-%d')
+    candidate_html = 'Canidate-{}-{}.html'.format(uuid.uuid4(), date)
+    context = {
+        'user_data' : user_data
+    }
+    utils.save_html(candidate_html, context, 'candidate.html')
+
+def main():
```

Comment:
This main function is really quite long. If it can be broken up into smaller functions, that would be good -- it is hard to read as it is.